api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -17,38 +16,84 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 option (xds.annotations.v3.file_status).work_in_progress = true;
 
 // [#protodoc-title: APIKey Auth]
-// [#not-implemented-hide:]
 // APIKey Auth :ref:`configuration overview <config_http_filters_api_key_auth>`.
 // [#extension: envoy.filters.http.api_key_auth]
 
 // API Key HTTP authentication.
 //
-// Example:
+// For example, the following configuration configures the filter to authenticate the clients using
+// the API key from the header ``X-API-KEY``. And only the clients with the key ``real-key`` are
+// considered as authenticated.
 //
 // .. code-block:: yaml
 //
-//    authentication_header: "X-API-KEY"
-//    keys:
-//      inline_string: |-
-//        clientID1:apiKey1
-//        clientID2:apiKey2
+//    credentials:
+//    - key: real-key
+//      client: user
+//    key_sources:
+//     - header: "X-API-KEY"
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+message ApiKeyAuth {
+  // The credentials that are used to authenticate the clients.
+  repeated Credential credentials = 1 [(udpa.annotations.sensitive) = true];
+
+  // The key sources to fetch the key from the coming request.
+  repeated KeySource key_sources = 2;
+}
+
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerRoute {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to **override** the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if non-empty ``credentials`` is set in the ``override_config``, then the
+  // ``credentials`` in the filter level configuration will be ignored and the ``credentials`` in
+  // this configuration will be used.
+  ApiKeyAuth override_config = 1;
+
+  // A list of clients that are allowed to access the route or vhost. The clients listed here
+  // should be subset of the clients listed in the ``credentials`` to provide authorization control
+  // after the authentication is successful. If the list is empty, then all authenticated clients
+  // are allowed. This provides very limited but simple authorization. If more complex authorization
+  // is required, then use the :ref:`HTTP RBAC filter <config_http_filters_rbac>` instead.
+  //
+  // .. note::
+  //   Setting this field and ``override_config.credentials`` at the same configuration entry is not
+  //   an error but also makes no sense because they provide similar functionality. Please only use
+  //   one of them at same configuration entry.
```

Comment:
Why wouldn't this be an error?
IIUC all the options here are essentially a oneof. As this is a newly introduced feature I think it can either reject or declare the order (prioritization) of these fields.

Reply:
The `authentication` and `authorization` are two different phases. Even both `credentials` and `allowed_clients` are configured, the filter could handle them correctly as expected.

So, I don't think this is an error and don't deserve a configuration rejection. The prioritization also only make the implementation more complex and make no much sense.